component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -320,27 +320,29 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
         String contextToken = (String) context.getProperty(EmailOTPAuthenticatorConstants.OTP_TOKEN);
         long generatedTime = (long) context.getProperty(EmailOTPAuthenticatorConstants.OTP_GENERATED_TIME);
         boolean isExpired = isExpired(generatedTime, context);
-
+        boolean succeededAttempt = false;
         if (userToken.equals(contextToken) && !isExpired) {
             processValidUserToken(context, authenticatedUser);
+            succeededAttempt = true;
         } else if (isBackupCodeEnabled(context)) {
-            validateWithBackUpCodes(context, userToken, authenticatedUser);
+            succeededAttempt = validateWithBackUpCodes(context, userToken, authenticatedUser);
         } else {
+            context.setProperty(EmailOTPAuthenticatorConstants.CODE_MISMATCH, true);
             if (isExpired) {
                 if (log.isDebugEnabled()) {
                     log.debug("Given otp code is expired.");
                 }
                 context.setProperty(EmailOTPAuthenticatorConstants.OTP_EXPIRED, "true");
-                handleOtpVerificationFail(context);
-                throw new AuthenticationFailedException("Code expired.");
             } else {
                 if (log.isDebugEnabled()) {
                     log.debug("Given otp code is mismatch.");
                 }
-                handleOtpVerificationFail(context);
-                throw new AuthenticationFailedException("Code mismatch.");
             }
         }
+        if (!succeededAttempt) {
+            handleOtpVerificationFail(context);
+            throw new AuthenticationFailedException("Code mismatch.");
```

Comment:
Is it possible to throw an error code here?

Reply:
There is no constructor in `AuthenticationFailedException` to pass an error code